File: 27. Longest Subarray with Sum K.py
# Topic: Longest Subarray with Sum K (handles negatives)

from collections import defaultdict

def longest_subarray_sum_k(nums, K):
    first = {0: -1}
    s = 0
    res = 0
    for i, v in enumerate(nums):
        s += v
        if (s - K) in first:
            res = max(res, i - first[s - K])
        if s not in first:
            first[s] = i
    return res

if __name__ == '__main__':
    nums = [1,-1,5,-2,3]
    K = 3
    print('Longest length:', longest_subarray_sum_k(nums,K))

# Time Complexity: O(n)
# Space Complexity: O(n)

File: 27. Longest Subarray with Sum K.cpp
// Topic: Longest Subarray with Sum K (handles negatives)
#include <bits/stdc++.h>
using namespace std;

int longestSubarraySumK(vector<int>& nums, int K)
{
    unordered_map<int, int> first;
    first[0] = -1;
    int sum = 0, res = 0;
    for (int i = 0; i < nums.size(); ++i)
    {
        sum += nums[i];
        if (first.find(sum - K) != first.end())
        {
            res = max(res, i - first[sum - K]);
        }
        if (first.find(sum) == first.end())
            first[sum] = i;
    }
    return res;
}

int main()
{
    vector<int> nums = {1, -1, 5, -2, 3};
    int K = 3;
    cout << "Longest length: " << longestSubarraySumK(nums, K) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)
